## Airbnb EDA
What are some questions we would like to ask if we had all the data about airbnb?
1. How many bookings happened in the dataset?
2. Which country had the most reviews/bookings?
3. What are the difference in the level of reviews amoungst different country?
4. What features are available that makes the airbnb more convenient
5. What are the effects of the security on the people?
6. What are the most common types of properties (e.g. shared room, private house, etc)?
7. What is the distribution of the number of listing one host has?
8. What is the statistical distribution of the pricing of listings?
9. Which time of the year have the highest and lowest pricing?
10. What are some of the keywords used in a review that has high or low review rating?

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('airbnb.csv')
pd.set_option('display.max_columns', None)

In [3]:
print(df.columns)
df.head()

Index(['Unnamed: 0', 'id', 'name', 'rating', 'reviews', 'host_name', 'host_id',
       'address', 'features', 'amenities', 'safety_rules', 'hourse_rules',
       'img_links', 'price', 'country', 'bathrooms', 'beds', 'guests',
       'toiles', 'bedrooms', 'studios', 'checkin', 'checkout'],
      dtype='object')


Unnamed: 0        id                                               name  \
0           0  49849504                                     Perla bungalov   
1           1  50891766  Authentic Beach Architect Sheltered Villa with...   
2           2  50699164                                  cottages sataplia   
3           3  49871422                        Sapanca Breathable Bungalow   
4           4  51245886                                      Bungalov Ev 2   

  rating reviews    host_name      host_id                   address  \
0   4.71      64    Mehmetcan  357334205.0  Kartepe, Kocaeli, Turkey   
1    New       0        Fatih  386223873.0      Kaş, Antalya, Turkey   
2   4.85      68       Giorgi  409690853.0          Imereti, Georgia   
3    5.0      13        Melih  401873242.0  Sapanca, Sakarya, Turkey   
4    New       0  Arp Sapanca  414884116.0  Sapanca, Sakarya, Turkey   

                                 features  \
0    2 guests,2 bedrooms,1 bed,1 bathroom   
1  4 guests,2 bedrooms,2 beds,2 bathrooms   
2    4 guests,1 bedroom,3 beds,1 bathroom   
3    4 guests,1 bedroom,2 beds,1 bathroom   
4     2 guests,1 bedroom,1 bed,1 bathroom   

                                           amenities  \
0  Mountain view,Valley view,Lake access,Kitchen,...   
1  Kitchen,Wifi,Dedicated workspace,Free parking ...   
2  Mountain view,Kitchen,Wifi,Dedicated workspace...   
3  Mountain view,Valley view,Kitchen,Wifi,Free pa...   
4  Kitchen,Wifi,Free parking on premises,TV,Air c...   

                                        safety_rules  \
0  󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...   
1  󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...   
2  󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...   
3  󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...   
4  󹀁,Airbnb's COVID-19 safety practices apply,󱠆,C...   

                                        hourse_rules  \
0  Check-in: Flexible,Check out: 12:00 pm,Pets ar...   
1  Check-in: 4:00 pm - 11:00 pm,Check out: 10:00 ...   
2  Check-in: After 1:00 pm,Check out: 12:00 pm,Se...   
3  Check-in: After 2:00 pm,Check out: 12:00 pm,No...   
4  Check-in: After 2:00 pm,Check out: 12:00 pm,No...   

                                           img_links  price   country  \
0  https://a0.muscache.com/im/pictures/a5da5cb7-c...   8078    Turkey   
1  https://a0.muscache.com/im/pictures/61b70855-2...   4665    Turkey   
2  https://a0.muscache.com/im/pictures/miso/Hosti...   5991   Georgia   
3  https://a0.muscache.com/im/pictures/72e6396e-e...  11339    Turkey   
4  https://a0.muscache.com/im/pictures/73973308-e...   6673    Turkey   

   bathrooms  beds  guests  toiles  bedrooms  studios              checkin  \
0          1     1       2       0         2        0             Flexible   
1          2     2       4       0         2        0   4 00 pm - 11 00 pm   
2          1     3       4       0         1        0        After 1 00 pm   
3          1     2       4       0         1        0        After 2 00 pm   
4          1     1       2       0         1        0        After 2 00 pm   

    checkout  
0   12 00 pm  
1   10 00 am  
2   12 00 pm  
3   12 00 pm  
4   12 00 pm

Columns are already formatted to allow indexing.

In [13]:
print(df.columns)

Index(['Unnamed: 0', 'id', 'name', 'rating', 'reviews', 'host_name', 'host_id',
       'address', 'features', 'amenities', 'safety_rules', 'hourse_rules',
       'img_links', 'price', 'country', 'bathrooms', 'beds', 'guests',
       'toiles', 'bedrooms', 'studios', 'checkin', 'checkout'],
      dtype='object')


### Columns to explore:
| Columns | Reason |
| ------- | ------ |
| Amenities | To help explore the relationship between amenities and rating/pricing/etc. |
| rating | Rating could help us evaluate the popularity of the listing |
| reviews | To balance the weight of the rating in respect to quantity of reviews |
| price | Help us explore the distribution of pricing across different countries |
| host_id | will allow us to see how many listings each host has |
| Country | see the distribution of listing prices, ratings, and reviews across different countries|
| safety_rules | relationship between safety rules and the liklihood of people rating the hosue high |
| house_rules | relationship between house rules and the liklihood of people rating house high |
| img_links | check if there are pictures in the listing/ connect to ratings |

### Columns to drop:
* Unnamed: 0
* host_name
* address 
* features

Amenities: To help explore the relationship between amenities and rating/pricing/etc.
rating: Rating could help us evaluate the popularity of the listing
reviews: To balance the weight of the rating in respect to quantity of reviews
price: Help us explore the distribution of pricing across different countries
host_id: will allow us to see how many listings each host has
Country: see the distribution of listing prices, ratings, and reviews across different countrie
safety_rules: relationship between safety rules and the liklihood of people rating the hosue high
house_rules: relationship between house rules and the liklihood of people rating house high
img_links: check if there are pictures in the listing/ connect to ratings

### Columns to drop:
* Unnamed: 0
* host_name
* address 
* features

In [10]:
df.shape

(12805, 23)

There are a total of 12805 records of listings on Airbnb.

## Explore Columns and Fix column types
___

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12805 entries, 0 to 12804
Data columns (total 23 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    12805 non-null  int64  
 1   id            12805 non-null  int64  
 2   name          12805 non-null  object 
 3   rating        12805 non-null  object 
 4   reviews       12805 non-null  object 
 5   host_name     12797 non-null  object 
 6   host_id       12805 non-null  float64
 7   address       12805 non-null  object 
 8   features      12805 non-null  object 
 9   amenities     12805 non-null  object 
 10  safety_rules  12805 non-null  object 
 11  hourse_rules  12805 non-null  object 
 12  img_links     12805 non-null  object 
 13  price         12805 non-null  int64  
 14  country       12805 non-null  object 
 15  bathrooms     12805 non-null  int64  
 16  beds          12805 non-null  int64  
 17  guests        12805 non-null  int64  
 18  toiles        12805 non-nu

* There are a couple of column names that are mispelled (need to fix)

In [4]:
df['checkout'].unique()

array([' 12 00 pm', ' 10 00 am', ' 11 00 am', nan, ' 3 00 pm', ' 1 00 pm',
       ' 12 00 am', ' 6 00 pm', ' 4 00 pm', ' 5 00 pm', ' 9 00 am',
       ' 2 00 pm', ' 8 00 am', ' 7 00 am', ' 1 00 am', ' 2 00 am',
       ' 11 00 pm', ' 10 00 pm', ' 3 00 am', ' 8 00 pm', ' 7 00 pm',
       ' 4 00 am'], dtype=object)

In [7]:
df[df['country'] == ' Japan']

Unnamed: 0                  id  \
39             39            21082489   
51             51  679805160588663838   
54             54            34029409   
1191         1191             3209511   
1204         1204            19256665   
...           ...                 ...   
12745       12745  585129042504781524   
12761       12761            49630137   
12788       12788            43681892   
12797       12797            44080105   
12804       12804            40547903   

                                                    name rating reviews  \
39                                  LOGIN OKINAWA -wood-   4.68     241   
51                                LOGIN OKINAWA -plants-    New       0   
54                     恩纳村的打卡圣地网红海景小木屋Cinderella J House   4.58      58   
1191                            Beach Side House alohana   4.99     249   
1204   A small house in the forest / Relaxing environ...   4.64     173   
...                                                  ...    ...     ...   
12745                URUMA DOME Earth4 & Mars5 MK309,310   4.67      12   
12761  Comfort Privacy and Nature with a View.  Pets OK!   4.86       7   
12788  Enjoy Vitamin SEA!  Ocean Front! [Ocean's Seven2]   4.95      21   
12797  Enjoy Vitamin SEA!  Ocean Front! [Ocean's Seve...    5.0       6   
12804   Enjoy Vitamin SEA!  Ocean Front! [Ocean's Seven]   4.85      13   

          host_name      host_id  \
39     LoginOkinawa  151975076.0   
51     LoginOkinawa  151975076.0   
54           Kiyomi  247159295.0   
1191      Miki&Masa   10312136.0   
1204        Daisuke   23602769.0   
...             ...          ...   
12745          Hiro  374017994.0   
12761      Stephane  400278563.0   
12788          Achy  246470880.0   
12797          Achy  246470880.0   
12804          Achy  246470880.0   

                                             address  \
39        Onna-son, Kunigami-gun, Okinawa-ken, Japan   
51           Onna, Kunigami District, Okinawa, Japan   
54                Onna, Kunigami-gun, Okinawa, Japan   
1191             Ishigaki, Okinawa Prefecture, Japan   
1204       Nishi-ku, Fukuoka-shi, Fukuoka-ken, Japan   
...                                              ...   
12745                          Uruma, Okinawa, Japan   
12761  Yuzawa, Minamiuonuma District, Niigata, Japan   
12788                          Uruma, Okinawa, Japan   
12797                          Uruma, Okinawa, Japan   
12804                          Uruma, Okinawa, Japan   

                                         features  \
39          9 guests,3 bedrooms,7 beds,1 bathroom   
51          8 guests,3 bedrooms,6 beds,1 bathroom   
54          8 guests,3 bedrooms,4 beds,1 bathroom   
1191         2 guests,1 bedroom,2 beds,1 bathroom   
1204         2 guests,1 bedroom,2 beds,1 bathroom   
...                                           ...   
12745        4 guests,1 bedroom,2 beds,1 bathroom   
12761  3 guests,1 bedroom,2 beds,Toilet with sink   
12788        4 guests,1 bedroom,2 beds,1 bathroom   
12797        4 guests,1 bedroom,3 beds,1 bathroom   
12804        5 guests,1 bedroom,2 beds,1 bathroom   

                                               amenities  \
39     Resort view,Ocean view,Beach access – Beachfro...   
51     Beach access,Kitchen,Wifi,Dedicated workspace,...   
54     Garden view,Beach access,Kitchen,Wifi,Dedicate...   
1191   Garden view,Beach access,Wifi,Dedicated worksp...   
1204   Courtyard view,Kitchen,Wifi,Free parking on pr...   
...                                                  ...   
12745  Waterfront,Kitchen,Wifi,Free parking on premis...   
12761  Mountain view,Public or shared beach access,Ki...   
12788  Ocean view,Sea view,Beach access – Beachfront,...   
12797  Ocean view,Sea view,Beach access,Kitchen,Wifi,...   
12804  Sea view,Beach access – Beachfront,Kitchen,Wif...   

                                            safety_rules  \
39     󹀁,Airbnb's COVID-19 safety practices apply,󱠃,N...   
51     󹀁,Airbnb's COVID-19 sa

## Summary of things To Do and fix
| Column | Number of non null values | Original Type | Type to Change to | To Do |
| ------ | ------------------------- | ------------- | ----------------- | ----- |
| Unnamed: 0    | 12805 |  int64   | | DROP |
| id            | 12805 |  int64   | |  |
| name          | 12805 |  object  | str | DROP after matching id and name |
| rating        | 12805 |  object  | str | * mainly float values, but when there are no rating, says "new" |
| reviews       | 12805 |  object  | int | * change type to int to make reviews easy to work with |
| host_name     | 12797 |  object  | | DROP after matching host name and host id |
| host_id       | 12805 |  float64 | int | |
| address       | 12805 |  object  | | DROP since we will be exploring country distribution instead of specific regions |
| features      | 12805 |  object  | | DROP since we have a summary of the features under other columns (toiles, bedrooms, studios, checkin, checkout) |
| amenities     | 12805 |  object  | | values are strung together with commas (,). when analyzing, we need to parse the strings |
| safety_rules  | 12805 |  object  | | values are strung together with commas (,). when analyzing, we need to parse the strings |
| hourse_rules  | 12805 |  object  | | * values are strung together with commas (,). when analyzing, we need to parse the strings </br> * rename column from hourse_rul to house_rule |
| hourse_rules  | 12805 |  object  | | * values are strung together with commas (,). when analyzing, we need to parse the strings </br> * rename column from hourse_rul to house_rule |
| img_links     | 12805 |  object  | | potentially use this column to see if the images help with reviews, but last in priority|
| price         | 12805 |  int64   | float | * Need to find out if the prices are recorded as US dollars  </br> * currently recroded as int, but investigate how it's being recorded|
| country       | 12805 |  object  | str | * Values have whitespaces, should remove |
| bathrooms     | 12805 |  int64   | | |
| beds          | 12805 |  int64   | | |
| guests        | 12805 |  int64   | | |
| toiles        | 12805 |  int64   | | * need to rename column|
| bedrooms      | 12805 |  int64   | | |
| studios       | 12805 |  int64   | | |
| checkin       | 12005 |  object  | | |
| checkout      | 10355 |  object  | datetime | * change to datetime </br> * figure out what to do with the null values|

Clean up the row values 

In [16]:
df.reviews.unique

<bound method Series.unique of 0        64
1         0
2        68
3        13
4         0
         ..
12800    16
12801     6
12802     3
12803     7
12804    13
Name: reviews, Length: 12805, dtype: object>

In [13]:
df[['country', 'rating', 'reviews', 'price']].head()

country rating reviews  price
0    Turkey   4.71      64   8078
1    Turkey    New       0   4665
2   Georgia   4.85      68   5991
3    Turkey    5.0      13  11339
4    Turkey    New       0   6673

In [39]:
country_rating = df[['country', 'rating']]
df.groupby('country').mean().drop(['Unnamed: 0', 'id','host_id'], axis = 1).head()

/var/folders/vh/dl6t7pz56gz05sm_y2j5fq840000gn/T/ipykernel_35161/852209497.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('country').mean().drop(['Unnamed: 0', 'id','host_id'], axis = 1).head()


price  bathrooms      beds    guests    toiles  bedrooms  \
country                                                                       
 Albania     4749.750000   3.500000  4.250000  6.750000  0.000000  3.500000   
 Argentina  10647.000000   1.333333  3.333333  5.333333  0.000000  1.333333   
 Armenia     6591.166667   0.833333  1.833333  3.166667  0.166667  1.000000   
 Australia  26735.000000   1.240000  2.640000  4.040000  0.000000  1.920000   
 Austria    20849.555556   1.746032  4.412698  5.746032  0.031746  2.365079   

             studios  
country               
 Albania    0.000000  
 Argentina  0.000000  
 Armenia    0.000000  
 Australia  0.000000  
 Austria    0.031746